### DISCUSSION TITLE
Question about inode invalidation when FUSE_CAP_WRITEBACK_CACHE is enabled

### DISCUSSION THREAD

**Discussion by apeters1971** [libfuse/libfuse#868] [General]
When I enable write-back cache with a 4.18 or 5.14 kernel, callbacks invalidating attributes of an inode e.g. a remote size change are not applied and the current size is stale. Is that a known issue or a limitation?

Is that due to this comment in the kernel source (in this case 5.14): 

	/*
	 * In case of writeback_cache enabled, the cached writes beyond EOF
	 * extend local i_size without keeping userspace server in sync. So,
	 * attr->size coming from server can be stale. We cannot trust it.
	 */

**Comment by Nikratio** [comment]
Could you clarify what you mean with `callbacks invalidating attributes of an inode`?

When writeback caching is enabled, it is expected that the FUSE userspace process does not have up-to-date information on e.g. inode size or mtime. However, it should eventually receive a `setattr` call from the kernel that provides the updated values.

**Comment by apeters1971** [comment]
Sure, 
I have a distributed filesystem use case. Consider two nodes A,B mounting a distributed filesystem. 

On node A I do:
`stat /fuse/file # let's say inode 1000`
On node B someone appends to inode 1000 
I receive a notification with the new size on node A and my fuse implementation calls:
```
 fuse_lowlevel_notify_inval_inode(session, 1000,0,0);
```
Without FUSE_CAP_WRITEBACK_CACHE the next stat on node A returns the new size notified from node B. While with FUSE_CAP_WRITEBACK_CACHE the next stat on node A returns the old size seen on node A.

**Comment by Nikratio** [comment]
I'd say that is the least surprising behaviour rather than a limitation or a bug. What do you think the kernel should do instead?

Imagine the kernel received a write request that increases the file size from x to x + 200. Because of writeback caching, this is not yet known to your FUSE process. Now you're instructing the kernel to "drop" this knowledge and (I assume) call getattr() to get updated values. But what if the updated size is less than x + 200, or the updated mtime is older than the write that the kernel has processed? Should it silently drop the extra data that it accepted? 

I guess another reasonable behavior would be to immediate start writeback and forget the values afterwards, but I wouldn't say that the current behavior is wrong...

**Reply by apeters1971** [reply]
I understand the behavior if node A has the file open for writing. But node A does nothing with the file, it just did a stat before and does not see the changes pushed from node B, just because A has the write back cache enabled and not in use for the file!

**Reply by ghz-max** [reply]
I faced the same kind of issue and I plan to broadcast a `truncate -s <size> <file>` to my fuse client/mount (using a message queue). This will update the size from the kernel point of view and invalidate the page cache of the file as well.

The command `echo 2 | sudo tee /proc/sys/vm/drop_caches` technically works and should only drop dentries and inodes but for fuse's files, it evicts the page cache as well. I wasn't able to find a Kernel API/syscall to forget a file in particular.
